File: projects/06/src/SymbolTable.cpp
#pragma once
#include "SymbolTable.h"

namespace HackAssembler
{
    SymbolTable::SymbolTable()
    {
        SymbolTable::symbol_table = {{"R0", 0}, {"R1", 1}, {"R2", 2}, {"R3", 3}, {"R4", 4}, {"R5", 5}, {"R6", 6},
            {"R7", 7}, {"R8", 8}, {"R9", 9}, {"R10", 10}, {"R11", 11}, {"R12", 12}, {"R13", 13}, {"R14", 14},
            {"R15", 15}, {"SP", 0}, {"LCL", 1}, {"ARG", 2}, {"THIS", 3}, {"THAT", 4}, {"SCREEN", 0x4000},
            {"KBD", 0x6000}};
    }

    SymbolTable::~SymbolTable()
    {
        SymbolTable::symbol_table.clear();
    }

    void SymbolTable::addEntry(const std::string& symbol, const uint16_t address)
    {
        if (address > 65535)
        {
            throw std::runtime_error(
                "Buffer overflow caused by symbol: " + symbol + "and an address value of: " + std::to_string(address));
        }
        SymbolTable::symbol_table.insert({symbol, address});
    }

    const bool SymbolTable::contains(const std::string& symbol) noexcept
    {
        return SymbolTable::symbol_table.find(symbol) != SymbolTable::symbol_table.cend();
    }

    const uint16_t SymbolTable::getAddress(const std::string& symbol)
    {
        return SymbolTable::symbol_table[symbol];
    }
} // namespace HackAssembler

File: projects/06/app/src/main/java/io/github/ironking24/nand2tetris/HackAssembler/SymbolTable.java
package io.github.ironking24.nand2tetris.HackAssembler;

import java.util.HashMap;

/**
 * Symbol reference handler, stores and resolves actual addresses of Symbols.
 * @author Ironking24
 * @version 1.0.0 
 */
public class SymbolTable {
    
    private final HashMap<String,Integer> definedLabels;

    SymbolTable(){
        definedLabels = new HashMap<>(21);
        definedLabels.put("R0", 0);
        definedLabels.put("R1", 1);
        definedLabels.put("R2", 2);
        definedLabels.put("R3", 3);
        definedLabels.put("R4", 4);
        definedLabels.put("R5", 5);
        definedLabels.put("R6", 6);
        definedLabels.put("R7", 7);
        definedLabels.put("R8", 8);
        definedLabels.put("R9", 9);
        definedLabels.put("R10", 10);
        definedLabels.put("R11", 11);
        definedLabels.put("R12", 12);
        definedLabels.put("R13", 13);
        definedLabels.put("R14", 14);
        definedLabels.put("R15", 15);
        definedLabels.put("SP", 0);
        definedLabels.put("LCL", 1);
        definedLabels.put("ARG", 2);
        definedLabels.put("THIS", 3);
        definedLabels.put("THAT", 4);
        definedLabels.put("SCREEN", 0x4000);
        definedLabels.put("KBD", 0x6000);
    }

    /**
     * Adds the symbol and it's corresponding memory address or file location to the table.
     * @param symbol  A string representation.
     * @param address  An integer that represents the address value.
    */
    public void addEntry(String symbol, int address){
        definedLabels.put(symbol, address);
    }

    /**
     * Checks if the symbol table has the given symbol.
     * @param symbol  The symbol to check for.
     * @return True if the table contains the symbol, false otherwise.
    */
    public boolean contains(String symbol){
        return definedLabels.containsKey(symbol);
    }

    /**
     * Returns the address value associated with the symbol.
     * @param symbol  The symbol to check for.
     * @return address value associated with the symbol.
     * @throws IllegalArgumentException If the symbol is unrecognized.
    */
    public int getAddress(String symbol) throws IllegalArgumentException{
        if(contains(symbol)){
            return definedLabels.get(symbol);
        }
        else{
            throw new IllegalArgumentException("The symbol table doesn't contain" + symbol);
        }
    }
}
